packages/devtools_app_shared/lib/src/service/service_extension_manager.dart

```diff
@@ -254,68 +254,79 @@ final class ServiceExtensionManager with DisposerMixin {
     }
     _hasServiceExtension(name).value = true;
 
-    if (_enabledServiceExtensions.containsKey(name)) {
+    final enabledServiceExtension = _enabledServiceExtensions[name];
+    if (enabledServiceExtension != null) {
       // Restore any previously enabled states by calling their service
       // extension. This will restore extension states on the device after a hot
       // restart. [_enabledServiceExtensions] will be empty on page refresh or
       // initial start.
       try {
-        return await _callServiceExtension(
+        final called = await _callServiceExtension(
           name,
-          _enabledServiceExtensions[name]!.value,
+          enabledServiceExtension.value,
         );
+        if (called) {
+          _serviceExtensions.add(name);
+        }
+        return;
       } on SentinelException catch (_) {
         // Service extension stopped existing while calling, so do nothing.
         // This typically happens during hot restarts.
       }
     } else {
       // Set any extensions that are already enabled on the device. This will
       // enable extension states in DevTools on page refresh or initial start.
-      return await _restoreExtensionFromDevice(name);
+      final restored = await _restoreExtensionFromDevice(name);
+      if (restored) {
+        _serviceExtensions.add(name);
+      }
     }
   }
 
   IsolateRef? get _mainIsolate => _isolateManager.mainIsolate.value;
 
-  Future<void> _restoreExtensionFromDevice(String name) async {
+  /// Restores the service extension named [name] from the device.
+  ///
+  /// Returns whether the service extension was actually restored.
+  Future<bool> _restoreExtensionFromDevice(String name) async {
     final isolateRef = _isolateManager.mainIsolate.value;
-    if (isolateRef == null) return;
+    if (isolateRef == null) return false;
 
     if (!extensions.serviceExtensionsAllowlist.containsKey(name)) {
-      return;
+      return false;
     }
     final expectedValueType =
         extensions.serviceExtensionsAllowlist[name]!.values.first.runtimeType;
 
-    Future<void> restore() async {
+    Future<bool> restore() async {
       // The restore request is obsolete if the isolate has changed.
-      if (isolateRef != _mainIsolate) return;
+      if (isolateRef != _mainIsolate) return false;
       try {
         final response = await _service!.callServiceExtension(
           name,
           isolateId: isolateRef.id,
         );
 
-        if (isolateRef != _mainIsolate) return;
+        if (isolateRef != _mainIsolate) return false;
 
         switch (expectedValueType) {
           case const (bool):
             final enabled = response.json!['enabled'] == 'true' ? true : false;
             await _maybeRestoreExtension(name, enabled);
```

Comment:
should `_maybeRestoreExtension` return a bool, and then we can `return await _maybeRestoreExtension`? Not sure if if matters that we don't always call the extension from `_maybeRestoreExtension`

Reply:
If we have `_maybeRestoreExtension` return a bool, then it will always return `true`, which I think would be silly, or it could delegate the return value to the two `setServiceExtensionState` calls it makes, but that function would also just always return `true`.